src/draggable-list/index.tsx

```diff
@@ -0,0 +1,177 @@
+	/** Called when an item is moved to a new position */
+	onMove: (from: number, to: number) => void;
+}
+
+export interface DraggableListChild {
+	content: DNode;
+	key: string;
+}
+
+interface DraggableListState {
+	dragIndex?: number;
+	itemHeight?: number;
+	itemWidth?: number;
+	keyRoot?: string;
+	lastPosition?: Position;
+	startPosition?: Position;
+	visualIndex?: number;
+}
+
+const factory = create({
+	dimensions,
+	drag,
+	icache: createICacheMiddleware<DraggableListState>(),
+	theme
+})
+	.properties<DraggableListProperties>()
+	.children<DraggableListChild[]>();
+
+export const DraggableList = factory(function DraggableList({
+	children,
+	middleware: { dimensions, drag, icache, theme },
+	properties
+}) {
+	const themeCss = theme.classes(css);
+	const items = children();
+
+	const { lockAxis, onMove } = properties();
+	const keyRoot = icache.get('keyRoot');
+	const lastPosition = icache.get('lastPosition');
+	const visualIndex = icache.get('visualIndex');
+	const renderIndex = visualIndex && Math.min(items.length - 1, Math.max(0, visualIndex));
+
+	let dragIndex = icache.get('dragIndex');
+	let itemHeight = icache.get('itemHeight');
+	let itemWidth = icache.get('itemWidth');
+	let startPosition = icache.get('startPosition');
+
+	const buildKey = (key: string) => `${keyRoot}-${key}`;
+
+	const renderItem = (item: DraggableListChild, index: number) => {
+		const itemDrag = drag.get(buildKey(item.key));
+		let currentPosition;
+		let padTop = false;
+		let padBottom = false;
+
+		// If this is a new drag, cache the starting item position, index, and height
+		if (itemDrag.start) {
+			const x = itemDrag.start.client.x - itemDrag.start.offset.x;
+			const y = itemDrag.start.client.y - itemDrag.start.offset.y;
+			startPosition = icache.set('startPosition', { x, y });
+			currentPosition = { x, y };
+			dragIndex = icache.set('dragIndex', index);
+			itemHeight = icache.set('itemHeight', dimensions.get(buildKey(item.key)).offset.height);
+			itemWidth = icache.set('itemWidth', dimensions.get(buildKey(item.key)).size.width);
+		}
+		// If this is an existing drag, use the last known position
+		else if (itemDrag.isDragging) {
+			currentPosition = lastPosition;
+		}
+
+		// Add top padding to the item immediately following the item being dragged
+		if (dragIndex !== undefined) {
+			if (
+				(renderIndex === undefined && index === dragIndex + 1) ||
+				(renderIndex !== undefined && renderIndex < dragIndex && index === renderIndex) ||
+				(renderIndex !== undefined && renderIndex >= dragIndex && index === renderIndex + 1)
+			) {
+				padTop = true;
+			}
+		}
+
+		// Add bottom padding to the end of the list if necessary
+		if (
+			renderIndex !== undefined &&
+			renderIndex === items.length - 1 &&
+			index === renderIndex
+		) {
+			padBottom = true;
+		}
+
+		// If this item is not being dragged, return a static list item
+		if (!itemDrag.isDragging || !startPosition || !currentPosition) {
+			return (
+				<li
```

Comment:
What benefit do we gain from adding the indirection of using a Dojo widget when no property or style is used from that widget, especially when we're using the `drag` middleware and doing complex DOM-based manipulation?  I feel like using raw HTML incurs less overhead internally and completely satisfies the needs of this widget.

Reply:
You'd still have your wrapping `li` domnode to which you'd apply the styling but would use the list item for it's contents as it would allow consistent styling / theming / active / selected / focused states etc